src/server/worker/spawner/common.go

```diff
@@ -1,31 +1,42 @@
 package spawner
 
-type Spawner interface {
-	RunService(Context.context) error
-	RunSpout(Context.context) error
-	RunMap(Context.context) error
-}
+func Run(pachClient *client.APIClient, pipelineInfo *pps.PipelineInfo, logger logs.TaggedLogger, utils common.Utils) error {
+	pipelineType, runFn := func() (string, SpawnerFunc) {
+		switch {
+		case pipelineInfo.Service != nil:
+			return "service", runService
+		case pipelineInfo.Spout != nil:
+			return "spout", runSpout
+		default:
+			return "pipeline", runMap
+		}
+	}()
 
-type spawner struct {
+	logger.Logf("Launching %v spawner process", pipelineType)
+	err := runFn(pachClient, pipelineInfo, logger, utils)
+	if err != nil {
+		logger.Logf("error running the %v spawner process: %v", pipelineType, err)
+	}
+	return err
 }
 
-func NewSpawner() Spawner {
+func runUserCode(
+	ctx context.Context,
+	logger logs.TaggedLogger,
+) error {
+	return runUntilCancel(ctx, "runUserCode", func() error {
+		// TODO: shouldn't this set up env like the worker does?
+		// TODO: what about the user error handling code?
+		return s.utils.RunUserCode(ctx, logger, nil, &pps.ProcessStats{}, nil)
+	})
 }
 
-func (s *spawner) runServiceUserCode(ctx context.Context, logger logs.TaggedLogger) error {
-	return backoff.RetryNotify(func() error {
-		// if we have a spout, then asynchronously receive spout data
-		if a.pipelineInfo.Spout != nil {
-			go a.receiveSpout(ctx, logger)
-		}
-		return a.runUserCode(ctx, logger, nil, &pps.ProcessStats{}, nil)
-	}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
-		select {
-		case <-ctx.Done():
+func runUntilCancel(ctx context.Context, name string, cb func() error) error {
+	return backoff.RetryNotify(cb, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
+		if common.IsDone(ctx) {
 			return err
-		default:
-			logger.Logf("error running user code: %+v, retrying in: %+v", err, d)
-			return nil
 		}
+		logger.Logf("error in %s: %+v, retrying in: %+v", name, err, d)
+		return nil
 	})
 }
```

Comment:
I think `runUntilCancel` is probably useful enough generally that it should just be migrated into the `backoff` package. Given how pervasive context is on our codebase I actually suspect that almost every usage of `Retry*` should exit when a context is cancelled. I'd call this `RetryUntilCancel` when it's located in `backoff`.